transport/internal/tls/dialer/dialer_test.go

```diff
@@ -0,0 +1,128 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dialer
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/transport/internal/tls/testscenario"
+)
+
+func TestDialer(t *testing.T) {
+	tests := []struct {
+		desc             string
+		withCustomDialer bool
+		data             string
+	}{
+		{desc: "without_custom_dialer", data: "test_no_dialer"},
+		{desc: "with_custom_dialer", data: "test_with_dialer", withCustomDialer: true},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			serverTLSConfig, clientTLSConfig := tlsConfigs(t)
+			params := Params{
+				Config: clientTLSConfig,
+			}
+			calledCustomDialer := false
+			if tt.withCustomDialer {
+				params.Dialer = func(ctx context.Context, network, address string) (net.Conn, error) {
+					calledCustomDialer = true
+					return (&net.Dialer{}).DialContext(ctx, network, address)
+				}
+			}
+			dialer := NewTLSDialer(params)
+			lis, err := net.Listen("tcp", "localhost:0")
+			require.NoError(t, err)
+			var wg sync.WaitGroup
+			defer wg.Wait()
+			defer lis.Close()
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				conn, err := lis.Accept()
+				require.NoError(t, err)
+				defer conn.Close()
+				tlsConn := tls.Server(conn, serverTLSConfig)
+
+				buf := make([]byte, len(tt.data))
+				n, err := tlsConn.Read(buf)
+				require.NoError(t, err)
+				_, err = tlsConn.Write(buf[:n])
+				assert.NoError(t, err)
+			}()
+
+			conn, err := dialer.DialContext(context.Background(), "tcp", lis.Addr().String())
+			require.NoError(t, err)
+
+			_, ok := conn.(*tls.Conn)
+			assert.True(t, ok)
+
+			n, err := conn.Write([]byte(tt.data))
+			require.NoError(t, err)
+			assert.Equal(t, len(tt.data), n)
+
+			buf := make([]byte, len(tt.data))
+			_, err = conn.Read(buf)
+			require.NoError(t, err)
+			assert.Equal(t, tt.data, string(buf))
+
+			if tt.withCustomDialer {
+				assert.True(t, calledCustomDialer)
+			}
+		})
+	}
+}
+
+func tlsConfigs(t *testing.T) (serverConfig *tls.Config, clientConfig *tls.Config) {
```

Comment:
oof I wish tls package provides some testing primitives so you don't have to do those.

Reply:
Yea I can move this method to tls/testscenario in a follow up

Reply:
#2176 done